File: android/src/main/java/com/plugin/admob/AdmobPlugin.kt
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

package com.plugin.admob

import android.app.Activity
import android.content.pm.PackageManager
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import app.tauri.annotation.Command
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import com.google.android.gms.ads.MobileAds

/** Log tag used for the plugin's own warnings and debug output. */
private const val TAG = "TauriAdmob"

/** Manifest meta-data key that must hold the app-level AdMob App ID. */
private const val MANIFEST_APP_ID_META_DATA = "com.google.android.gms.ads.APPLICATION_ID"

/**
 * Tauri v2 mobile plugin entry point for Google AdMob on Android.
 *
 * Registered from the Rust core via
 * `register_android_plugin("com.plugin.admob", "AdmobPlugin")`. Every command
 * hops to the UI thread before touching the AdMob SDK, the view hierarchy or
 * the event channel; `resolve`/`reject` happen exactly once per command.
 */
@TauriPlugin
class AdmobPlugin(private val activity: Activity) : Plugin(activity) {

    private val consentManager = ConsentManager(activity)
    private val bannerManager = BannerManager(activity) { event, payload -> trigger(event, payload) }
    private val interstitialManager =
        InterstitialManager(activity) { event, payload -> trigger(event, payload) }
    private val rewardedManager =
        RewardedManager(activity) { event, payload -> trigger(event, payload) }

    /** Set by the first successful `initialize`; lets a repeated call no-op. */
    private var initialized = false

    /** Enables verbose native logging (never logs personal data). */
    private var isDebug = false

    /** Last consent status reported through `initialize` / `requestConsent`. */
    private var lastConsentStatus = CONSENT_STATUS_UNKNOWN

    // ------------------------------------------------------------------
    // Commands
    // ------------------------------------------------------------------

    /**
     * Validates the manifest AdMob App ID, starts the Mobile Ads SDK, runs
     * the UMP consent flow when requested, triggers `admob://initialized` and
     * resolves `{ consentStatus, trackingStatus }`.
     */
    @Command
    fun initialize(invoke: Invoke) {
        val args = invoke.parseArgs(InitializeArgs::class.java)
        activity.runOnUiThread { handleInitialize(args, invoke) }
    }

    /** Runs the UMP consent flow and resolves `{ status }`. */
    @Command
    fun requestConsent(invoke: Invoke) {
        activity.runOnUiThread {
            consentManager.request { consentStatus, formError ->
                if (formError != null) {
                    Log.w(TAG, "consent request failed: ${formError.message}")
                    invoke.reject(
                        "consent request failed: ${formError.message}",
                        AdmobRejectCodes.NATIVE_ERROR
                    )
                } else {
                    val status = mapConsentStatus(consentStatus)
                    lastConsentStatus = status
                    trigger(AdmobEvents.CONSENT_CHANGED, JSObject().put("status", status))
                    invoke.resolve(JSObject().put("status", status))
                }
            }
        }
    }

    /**
     * Android has no App Tracking Transparency equivalent: resolves
     * `{ status: "notAvailable" }` without triggering any event.
     */
    @Command
    fun requestTrackingAuthorization(invoke: Invoke) {
        activity.runOnUiThread {
            invoke.resolve(JSObject().put("status", TRACKING_STATUS_NOT_AVAILABLE))
        }
    }

    /** Loads a fresh interstitial for `adUnitId`. */
    @Command
    fun loadInterstitial(invoke: Invoke) {
        val args = invoke.parseArgs(LoadAdArgs::class.java)
        val adUnitId = args.adUnitId
        if (adUnitId.isNullOrBlank()) {
            invoke.reject(
                "loadInterstitial requires a non-empty adUnitId argument",
                AdmobRejectCodes.INVALID_ARGUMENT
            )
            return
        }
        debugLog("loading interstitial")
        activity.runOnUiThread { interstitialManager.load(adUnitId, invoke) }
    }

    /** Shows the stored interstitial; resolves once it is dismissed. */
    @Command
    fun showInterstitial(invoke: Invoke) {
        activity.runOnUiThread { interstitialManager.show(invoke) }
    }

    /** Clears the stored interstitial and listeners. */
    @Command
    fun destroyInterstitial(invoke: Invoke) {
        activity.runOnUiThread {
            interstitialManager.destroy()
            invoke.resolve()
        }
    }

    /** Loads a fresh rewarded ad for `adUnitId`. */
    @Command
    fun loadRewarded(invoke: Invoke) {
        val args = invoke.parseArgs(LoadAdArgs::class.java)
        val adUnitId = args.adUnitId
        if (adUnitId.isNullOrBlank()) {
            invoke.reject(
                "loadRewarded requires a non-empty adUnitId argument",
                AdmobRejectCodes.INVALID_ARGUMENT
            )
            return
        }
        debugLog("loading rewarded ad")
        activity.runOnUiThread { rewardedManager.load(adUnitId, invoke) }
    }

    /** Shows the stored rewarded ad; resolves once it is dismissed. */
    @Command
    fun showRewarded(invoke: Invoke) {
        activity.runOnUiThread { rewardedManager.show(invoke) }
    }

    /** Clears the stored rewarded ad and listeners. */
    @Command
    fun destroyRewarded(invoke: Invoke) {
        activity.runOnUiThread {
            rewardedManager.destroy()
            invoke.resolve()
        }
    }

    /** Creates, attaches and loads the banner; resolves once the ad loads. */
    @Command
    fun showBanner(invoke: Invoke) {
        val args = invoke.parseArgs(ShowBannerArgs::class.java)
        val adUnitId = args.adUnitId
        if (adUnitId.isNullOrBlank()) {
            invoke.reject(
                "showBanner requires a non-empty adUnitId argument",
                AdmobRejectCodes.INVALID_ARGUMENT
            )
            return
        }
        val position = parseBannerPosition(args.position)
        if (position == null) {
            invoke.reject(
                "invalid banner position '${args.position}' (expected \"top\" or \"bottom\")",
                AdmobRejectCodes.INVALID_ARGUMENT
            )
            return
        }
        val size = parseBannerSize(args.size)
        if (size == null) {
            invoke.reject(
                "invalid banner size '${args.size}' (expected banner, largeBanner, " +
                    "mediumRectangle, fullBanner, leaderboard or adaptive)",
                AdmobRejectCodes.INVALID_ARGUMENT
            )
            return
        }
        debugLog("showing banner ($position, $size)")
        activity.runOnUiThread { bannerManager.show(adUnitId, position, size, invoke) }
    }

    /** Hides the banner, keeping the view alive for a re-show. */
    @Command
    fun hideBanner(invoke: Invoke) {
        activity.runOnUiThread {
            bannerManager.hide()
            invoke.resolve()
        }
    }

    /** Repositions the live banner; no-op when there is none. */
    @Command
    fun setBannerPosition(invoke: Invoke) {
        val args = invoke.parseArgs(SetBannerPositionArgs::class.java)
        val position = parseBannerPosition(args.position)
        if (position == null) {
            invoke.reject(
                "invalid banner position '${args.position}' (expected \"top\" or \"bottom\")",
                AdmobRejectCodes.INVALID_ARGUMENT
            )
            return
        }
        activity.runOnUiThread {
            bannerManager.setPosition(position)
            invoke.resolve()
        }
    }

    /** Removes and destroys the banner view, clearing every reference. */
    @Command
    fun destroyBanner(invoke: Invoke) {
        activity.runOnUiThread {
            bannerManager.destroy()
            invoke.resolve()
        }
    }

    /** Tears down banner, interstitial and rewarded state. */
    @Command
    fun destroy(invoke: Invoke) {
        activity.runOnUiThread {
            debugLog("destroying AdMob resources")
            bannerManager.destroy()
            interstitialManager.destroy()
            rewardedManager.destroy()
            initialized = false
            lastConsentStatus = CONSENT_STATUS_UNKNOWN
            invoke.resolve()
        }
    }

    // ------------------------------------------------------------------
    // Lifecycle
    // ------------------------------------------------------------------

    /**
     * The AdMob SDK keeps banner views current across foreground transitions
     * by itself (recent SDK versions pause/resume `AdView` internally), so
     * `onResume`/`onPause` are intentionally empty.
     */
    override fun onResume() {
    }

    /** See [onResume]: the SDK handles pause/resume internally. */
    override fun onPause() {
    }

    /** Releases the banner and the full-screen ads to avoid leaks. */
    override fun onDestroy(activity: AppCompatActivity) {
        activity.runOnUiThread {
            bannerManager.destroy()
            interstitialManager.destroy()
            rewardedManager.destroy()
            initialized = false
        }
    }

    // ------------------------------------------------------------------
    // Internals
    // ------------------------------------------------------------------

    private fun handleInitialize(args: InitializeArgs, invoke: Invoke) {
        if (initialized) {
            debugLog("initialize called again; already initialized")
            invoke.resolve(initializeResult(lastConsentStatus))
            return
        }

        if (args.appId.isNullOrBlank()) {
            invoke.reject(
                "initialize requires a non-empty appId argument",
                AdmobRejectCodes.INVALID_ARGUMENT
            )
            return
        }

        isDebug = args.debug

        val manifestAppId = manifestAdMobAppId()
        if (manifestAppId.isNullOrBlank()) {
            invoke.reject(
                "Missing AdMob App ID. Add " +
                    "<meta-data android:name=\"com.google.android.gms.ads.APPLICATION_ID\" " +
                    "android:value=\"ca-app-pub-xxxxx~xxxxx\"/> inside the <application> " +
                    "element of AndroidManifest.xml.",
                AdmobRejectCodes.INVALID_CONFIGURATION
            )
            return
        }
        if (manifestAppId != args.appId) {
            // The AdMob SDK always uses the manifest value; warn and continue.
            Log.w(
                TAG,
                "initialize appId '${args.appId}' differs from the manifest value " +
                    "'$manifestAppId'; the AdMob SDK always uses the manifest value"
            )
        }

        debugLog(
            "initializing MobileAds (testing=${args.isTesting}, " +
                "automaticallyRequestConsent=${args.automaticallyRequestConsent})"
        )

        try {
            MobileAds.initialize(activity) {
                if (args.automaticallyRequestConsent) {
                    consentManager.request { consentStatus, formError ->
                        if (formError != null) {
                            Log.w(TAG, "consent flow failed: ${formError.message}")
                            invoke.reject(
                                "consent flow failed: ${formError.message}",
                                AdmobRejectCodes.NATIVE_ERROR
                            )
                        } else {
                            val status = mapConsentStatus(consentStatus)
                            lastConsentStatus = status
                            trigger(
                                AdmobEvents.CONSENT_CHANGED,
                                JSObject().put("status", status)
                            )
                            finishInitialize(args, status, invoke)
                        }
                    }
                } else {
                    val status = mapConsentStatus(consentManager.currentStatus())
                    lastConsentStatus = status
                    finishInitialize(args, status, invoke)
                }
            }
        } catch (error: Exception) {
            invoke.reject(
                "MobileAds initialization failed: ${error.message}",
                AdmobRejectCodes.NATIVE_ERROR
            )
        }
    }

    /**
     * Triggers `admob://initialized` and resolves initialize. On Android
     * `trackingStatus` is always `notAvailable` (no ATT); the tracking event
     * is never triggered from this platform.
     */
    private fun finishInitialize(args: InitializeArgs, consentStatus: String, invoke: Invoke) {
        initialized = true
        trigger(
            AdmobEvents.INITIALIZED,
            JSObject()
                .put("platform", PLATFORM_ANDROID)
                .put("testing", args.isTesting)
        )
        invoke.resolve(initializeResult(consentStatus))
    }

    private fun initializeResult(consentStatus: String): JSObject = JSObject()
        .put("consentStatus", consentStatus)
        .put("trackingStatus", TRACKING_STATUS_NOT_AVAILABLE)

    /**
     * Reads the app-level AdMob App ID from the merged manifest meta-data.
     * Uses the classic `GET_META_DATA` lookup, which works on every supported
     * minSdk (the API 33+ overload is unavailable below Tiramisu).
     */
    private fun manifestAdMobAppId(): String? {
        return try {
            @Suppress("DEPRECATION")
            val appInfo = activity.packageManager.getApplicationInfo(
                activity.packageName,
                PackageManager.GET_META_DATA
            )
            appInfo.metaData?.getString(MANIFEST_APP_ID_META_DATA)
        } catch (_: PackageManager.NameNotFoundException) {
            null
        }
    }

    /** Debug-gated log helper; never called with personal data. */
    private fun debugLog(message: String) {
        if (isDebug) {
            Log.d(TAG, message)
        }
    }
}

File: ios/Sources/AdmobPlugin.swift
// Copyright 2026 tauri-plugin-ad2mob contributors
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

import Foundation
import GoogleMobileAds
import Tauri
import UIKit
import WebKit

// Argument payloads. Fields are optional so that malformed or partial
// payloads are reported as stable `INVALID_ARGUMENT` rejections instead of
// opaque `Decodable` failures escaping to the IPC layer.

/// `initialize` arguments (contract: `{ appId, isTesting, automaticallyRequestConsent, requestTrackingAuthorization, debug }`).
struct InitializeArgs: Decodable {
  let appId: String?
  let isTesting: Bool?
  let automaticallyRequestConsent: Bool?
  let requestTrackingAuthorization: Bool?
  let debug: Bool?
}

/// `loadInterstitial` / `loadRewarded` arguments (`{ adUnitId }`).
struct LoadAdArgs: Decodable {
  let adUnitId: String?
}

/// `showBanner` arguments (`{ adUnitId, position, size }`).
struct ShowBannerArgs: Decodable {
  let adUnitId: String?
  let position: String?
  let size: String?
}

/// `setBannerPosition` arguments (`{ position }`).
struct SetBannerPositionArgs: Decodable {
  let position: String?
}

/// Native entry point of the AdMob plugin. The Rust core registers it through
/// `register_ios_plugin(init_plugin_admob)` (see the `@_cdecl` export below).
///
/// The Tauri plugin manager dispatches commands on its background IPC queue
/// (`PluginManager.ipcDispatchQueue`), so EVERY command hops to the main
/// queue before touching the Google Mobile Ads SDK, UIKit or plugin state
/// (contract threading rules). Rejections only ever carry `{ code, message }`
/// and resolve/reject happen exactly once per invoke.
class AdmobPlugin: Plugin {
  /// Info.plist key holding the AdMob App ID (contract: app-owned configuration).
  private static let appIdentifierKey = "GADApplicationIdentifier"

  private let consentManager = ConsentManager()
  private let trackingManager = TrackingManager()
  private lazy var bannerManager = BannerManager(plugin: self)
  private lazy var interstitialManager = InterstitialManager(plugin: self)
  private lazy var rewardedManager = RewardedManager(plugin: self)

  /// Weak on purpose: the managers are owned by the plugin and must never
  /// keep it (or the WebView) alive.
  private weak var webView: WKWebView?

  // Main-queue-confined state: every command hops to the main queue before
  // reading or writing any of these.
  private var initialized = false
  private var initializationInFlight = false
  private var testing = false
  private var debugEnabled = false

  /// Called by the Tauri plugin manager when the WebView is created; captures
  /// the WebView as a fallback anchor for the banner container view.
  public override func load(webview: WKWebView) {
    self.webView = webview
  }

  // MARK: - Commands

  /// Starts the Mobile Ads SDK and (optionally) the consent and ATT flows.
  /// Resolves `{ consentStatus, trackingStatus }` with normalized values.
  @objc public func initialize(_ invoke: Invoke) throws {
    guard let args = decodeArgs(InitializeArgs.self, command: "initialize", invoke: invoke) else {
      return
    }
    guard let appId = args.appId, !appId.isEmpty else {
      invoke.reject("initialize requires a non-empty `appId`", code: "INVALID_ARGUMENT")
      return
    }

    DispatchQueue.main.async { [weak self] in
      guard let self else {
        invoke.reject("plugin deallocated", code: "NATIVE_ERROR")
        return
      }
      self.runInitialize(appId: appId, args: args, invoke: invoke)
    }
  }

  /// Runs the UMP consent flow on demand.
  /// Resolves `{ status }` with the normalized consent status.
  @objc public func requestConsent(_ invoke: Invoke) throws {
    DispatchQueue.main.async { [weak self] in
      guard let self else {
        invoke.reject("plugin deallocated", code: "NATIVE_ERROR")
        return
      }
      guard let viewController = self.rootViewController() else {
        invoke.reject(
          "no root view controller available to present the consent form", code: "NATIVE_ERROR")
        return
      }
      self.consentManager.requestConsent(from: viewController) { result in
        // ConsentManager hops to the main queue before calling back.
        switch result {
        case .success(let status):
          self.trigger(AdmobEvents.CONSENT_CHANGED, data: ["status": status])
          invoke.resolve(["status": status])
        case .failure(let flowError):
          invoke.reject(flowError.message, code: "NATIVE_ERROR")
        }
      }
    }
  }

  /// Runs the ATT prompt on demand.
  /// Resolves `{ status }` with the normalized authorization status
  /// (`"notAvailable"` below iOS 14).
  @objc public func requestTrackingAuthorization(_ invoke: Invoke) throws {
    DispatchQueue.main.async { [weak self] in
      guard let self else {
        invoke.reject("plugin deallocated", code: "NATIVE_ERROR")
        return
      }
      // ATT prompts are only shown while the app is active: requesting from
      // the main queue (we are on it here) satisfies that requirement.
      self.trackingManager.requestAuthorization { status in
        // TrackingManager hops to the main queue before calling back.
        self.trigger(AdmobEvents.TRACKING_AUTHORIZATION_CHANGED, data: ["status": status])
        invoke.resolve(["status": status])
      }
    }
  }

  /// Loads a fresh interstitial. Resolves once the SDK reports the ad ready.
  @objc public func loadInterstitial(_ invoke: Invoke) throws {
    guard let args = decodeArgs(LoadAdArgs.self, command: "loadInterstitial", invoke: invoke)
    else {
      return
    }
    guard let adUnitId = args.adUnitId, !adUnitId.isEmpty else {
      invoke.reject("loadInterstitial requires a non-empty `adUnitId`", code: "INVALID_ARGUMENT")
      return
    }
    DispatchQueue.main.async { [weak self] in
      guard let self else {
        invoke.reject("plugin deallocated", code: "NATIVE_ERROR")
        return
      }
      guard self.initialized else {
        invoke.reject("the plugin is not initialized; call initialize first", code: "NOT_INITIALIZED")
        return
      }
      self.interstitialManager.load(adUnitId: adUnitId, invoke: invoke)
    }
  }

  /// Presents the stored interstitial; resolves when the user dismisses it.
  @objc public func showInterstitial(_ invoke: Invoke) throws {
    DispatchQueue.main.async { [weak self] in
      guard let self else {
        invoke.reject("plugin deallocated", code: "NATIVE_ERROR")
        return
      }
      guard self.initialized else {
        invoke.reject("the plugin is not initialized; call initialize first", code: "NOT_INITIALIZED")
        return
      }
      self.interstitialManager.show(invoke: invoke)
    }
  }

  /// Clears the stored interstitial. Safe when nothing was loaded.
  @objc public func destroyInterstitial(_ invoke: Invoke) throws {
    DispatchQueue.main.async { [weak self] in
      self?.interstitialManager.destroy()
      invoke.resolve()
    }
  }

  /// Loads a fresh rewarded ad. Resolves once the SDK reports the ad ready.
  @objc public func loadRewarded(_ invoke: Invoke) throws {
    guard let args = decodeArgs(LoadAdArgs.self, command: "loadRewarded", invoke: invoke) else {
      return
    }
    guard let adUnitId = args.adUnitId, !adUnitId.isEmpty else {
      invoke.reject("loadRewarded requires a non-empty `adUnitId`", code: "INVALID_ARGUMENT")
      return
    }
    DispatchQueue.main.async { [weak self] in
      guard let self else {
        invoke.reject("plugin deallocated", code: "NATIVE_ERROR")
        return
      }
      guard self.initialized else {
        invoke.reject("the plugin is not initialized; call initialize first", code: "NOT_INITIALIZED")
        return
      }
      self.rewardedManager.load(adUnitId: adUnitId, invoke: invoke)
    }
  }

  /// Presents the stored rewarded ad; resolves when the user dismisses it.
  /// Rewards arrive through the `admob://rewarded-earned` event only.
  @objc public func showRewarded(_ invoke: Invoke) throws {
    DispatchQueue.main.async { [weak self] in
      guard let self else {
        invoke.reject("plugin deallocated", code: "NATIVE_ERROR")
        return
      }
      guard self.initialized else {
        invoke.reject("the plugin is not initialized; call initialize first", code: "NOT_INITIALIZED")
        return
      }
      self.rewardedManager.show(invoke: invoke)
    }
  }

  /// Clears the stored rewarded ad. Safe when nothing was loaded.
  @objc public func destroyRewarded(_ invoke: Invoke) throws {
    DispatchQueue.main.async { [weak self] in
      self?.rewardedManager.destroy()
      invoke.resolve()
    }
  }

  /// Creates and loads a banner attached to the Tauri root view.
  /// Resolves once the banner loaded and was revealed, rejects `LOAD_FAILED`
  /// otherwise (nothing stays attached after a failure).
  @objc public func showBanner(_ invoke: Invoke) throws {
    guard let args = decodeArgs(ShowBannerArgs.self, command: "showBanner", invoke: invoke) else {
      return
    }
    guard let adUnitId = args.adUnitId, !adUnitId.isEmpty else {
      invoke.reject("showBanner requires a non-empty `adUnitId`", code: "INVALID_ARGUMENT")
      return
    }
    guard let rawPosition = args.position,
      let parsedPosition = AdmobMappers.parseBannerPosition(rawPosition)
    else {
      invoke.reject(
        "showBanner requires `position` set to \"top\" or \"bottom\"", code: "INVALID_ARGUMENT")
      return
    }
    guard let rawSize = args.size, let parsedSize = AdmobMappers.parseBannerSize(rawSize) else {
      invoke.reject(
        "showBanner requires `size` set to \"banner\", \"largeBanner\", \"mediumRectangle\", "
          + "\"fullBanner\", \"leaderboard\" or \"adaptive\"",
        code: "INVALID_ARGUMENT")
      return
    }
    DispatchQueue.main.async { [weak self] in
      guard let self else {
        invoke.reject("plugin deallocated", code: "NATIVE_ERROR")
        return
      }
      guard self.initialized else {
        invoke.reject("the plugin is not initialized; call initialize first", code: "NOT_INITIALIZED")
        return
      }
      self.bannerManager.show(
        adUnitId: adUnitId, position: parsedPosition, size: parsedSize, invoke: invoke)
    }
  }

  /// Hides the banner view (kept for a later re-show). Safe when there is none.
  @objc public func hideBanner(_ invoke: Invoke) throws {
    DispatchQueue.main.async { [weak self] in
      self?.bannerManager.hide()
      invoke.resolve()
    }
  }

  /// Repositions a live banner (no-op when there is none).
  @objc public func setBannerPosition(_ invoke: Invoke) throws {
    guard let args = decodeArgs(SetBannerPositionArgs.self, command: "setBannerPosition", invoke: invoke)
    else {
      return
    }
    guard let position = args.position,
      AdmobMappers.parseBannerPosition(position) != nil
    else {
      invoke.reject(
        "setBannerPosition requires `position` set to \"top\" or \"bottom\"",
        code: "INVALID_ARGUMENT")
      return
    }
    DispatchQueue.main.async { [weak self] in
      self?.bannerManager.setPosition(position)
      invoke.resolve()
    }
  }

  /// Removes the banner from the view hierarchy and clears its references.
  /// Safe when there is none.
  @objc public func destroyBanner(_ invoke: Invoke) throws {
    DispatchQueue.main.async { [weak self] in
      self?.bannerManager.destroy()
      invoke.resolve()
    }
  }

  /// Tears down banner, interstitial and rewarded ads. Must be safe when
  /// nothing was initialized.
  @objc public func destroy(_ invoke: Invoke) throws {
    DispatchQueue.main.async { [weak self] in
      guard let self else {
        invoke.resolve()
        return
      }
      self.bannerManager.destroy()
      self.interstitialManager.destroy()
      self.rewardedManager.destroy()
      // Mirrors the Rust core: destroy resets the initialization state so the
      // plugin can be initialized again.
      self.initialized = false
      self.initializationInFlight = false
      self.testing = false
      self.debugEnabled = false
      invoke.resolve()
    }
  }

  // MARK: - Initialization flow

  /// Contract `initialize` flow. Runs on the main queue.
  private func runInitialize(appId: String, args: InitializeArgs, invoke: Invoke) {
    // Idempotency mirrors the Rust core: a repeated initialize reports the
    // current status without re-running the SDK flows.
    if initialized {
      invoke.resolve([
        "consentStatus": ConsentManager.currentStatus(),
        "trackingStatus": trackingManager.currentStatus(),
      ])
      return
    }
    if initializationInFlight {
      invoke.reject("an initialize call is already in progress", code: "NATIVE_ERROR")
      return
    }

    // Contract step 2: the app-level AdMob App ID is owned by the consuming
    // app and read from Info.plist.
    guard
      let plistAppId = Bundle.main.object(forInfoDictionaryKey: Self.appIdentifierKey) as? String,
      !plistAppId.isEmpty
    else {
      invoke.reject(
        "missing \(Self.appIdentifierKey) in Info.plist; the Google Mobile Ads SDK aborts "
          + "without it. Add <key>\(Self.appIdentifierKey)</key>"
          + "<string>ca-app-pub-XXXXXXXXXXXXXXXX~YYYYYYYYYY</string> to the app target's "
          + "Info.plist",
        code: "INVALID_CONFIGURATION")
      return
    }
    if plistAppId != appId {
      // The SDK always uses the Info.plist value; keep going with a warning
      // (common in development, where isTesting resolves the Google test App ID).
      Logger.info(
        "WARNING: AdMob appId argument (\(appId)) differs from Info.plist "
          + "\(Self.appIdentifierKey) (\(plistAppId)); the SDK uses the Info.plist value",
        category: "admob")
    }

    debugEnabled = args.debug ?? false
    testing = args.isTesting ?? false
    let wantsConsent = args.automaticallyRequestConsent ?? false
    let wantsTracking = args.requestTrackingAuthorization ?? false
    initializationInFlight = true

    logDebug("starting Google Mobile Ads SDK (testing: \(testing))")
    // Contract step 3: start the SDK. We are already on the main queue; the
    // completion may arrive elsewhere, so it hops back before continuing.
    GADMobileAds.sharedInstance().start { [weak self] _ in
      DispatchQueue.main.async {
        guard let self else {
          invoke.reject("plugin deallocated", code: "NATIVE_ERROR")
          return
        }
        self.continueAfterSdkStart(wantsConsent: wantsConsent, wantsTracking: wantsTracking, invoke: invoke)
      }
    }
  }

  /// Continues `initialize` after the SDK started: consent flow first.
  private func continueAfterSdkStart(wantsConsent: Bool, wantsTracking: Bool, invoke: Invoke) {
    if wantsConsent {
      guard let viewController = rootViewController() else {
        initializationInFlight = false
        invoke.reject(
          "no root view controller available to present the consent form", code: "NATIVE_ERROR")
        return
      }
      consentManager.requestConsent(from: viewController) { [weak self] result in
        // ConsentManager hops to the main queue before calling back.
        guard let self else {
          invoke.reject("plugin deallocated", code: "NATIVE_ERROR")
          return
        }
        switch result {
        case .success(let status):
          self.trigger(AdmobEvents.CONSENT_CHANGED, data: ["status": status])
          self.requestTrackingThenFinish(wantsTracking: wantsTracking, invoke: invoke)
        case .failure(let flowError):
          self.initializationInFlight = false
          invoke.reject(flowError.message, code: "NATIVE_ERROR")
        }
      }
    } else {
      requestTrackingThenFinish(wantsTracking: wantsTracking, invoke: invoke)
    }
  }

  /// Continues `initialize` after consent: ATT request when asked for.
  private func requestTrackingThenFinish(wantsTracking: Bool, invoke: Invoke) {
    if wantsTracking {
      // ATT must be requested while the app is active; chaining it here (main
      // queue, inside the start() completion chain) guarantees that, and the
      // prompt shows after the consent form (if any) was dismissed.
      trackingManager.requestAuthorization { [weak self] status in
        // TrackingManager hops to the main queue before calling back.
        guard let self else {
          invoke.reject("plugin deallocated", code: "NATIVE_ERROR")
          return
        }
        self.trigger(AdmobEvents.TRACKING_AUTHORIZATION_CHANGED, data: ["status": status])
        self.completeInitialization(trackingStatus: status, invoke: invoke)
      }
    } else {
      completeInitialization(trackingStatus: trackingManager.currentStatus(), invoke: invoke)
    }
  }

  /// Contract steps 5-6: trigger `admob://initialized`, resolve the status map.
  private func completeInitialization(trackingStatus: String, invoke: Invoke) {
    initializationInFlight = false
    initialized = true
    let consentStatus = ConsentManager.currentStatus()
    trigger(AdmobEvents.INITIALIZED, data: ["platform": "ios", "testing": testing])
    logDebug("initialized (consent: \(consentStatus), tracking: \(trackingStatus))")
    invoke.resolve(["consentStatus": consentStatus, "trackingStatus": trackingStatus])
  }

  // MARK: - Helpers

  /// Decodes the command arguments, rejecting with a stable `INVALID_ARGUMENT`
  /// code on failure. Returns nil after rejecting; callers must bail out.
  private func decodeArgs<T: Decodable>(_ type: T.Type, command: String, invoke: Invoke) -> T? {
    do {
      return try invoke.parseArgs(type)
    } catch {
      invoke.reject("invalid arguments for \(command): \(error)", code: "INVALID_ARGUMENT")
      return nil
    }
  }

  /// Root view controller used to present full-screen ads and consent forms:
  /// the Tauri plugin manager's view controller, falling back to the root of
  /// the window hosting the WebView.
  func rootViewController() -> UIViewController? {
    if let viewController = manager.viewController {
      return viewController
    }
    return webView?.window?.rootViewController
  }

  /// View the banner is attached to: the Tauri root view, falling back to the
  /// WebView's superview (captured in `load(webview:)`).
  func containerView() -> UIView? {
    if let view = manager.viewController?.view {
      return view
    }
    return webView?.superview
  }

  /// Logs a verbose message when `debug` was enabled through `initialize`.
  /// Never logs personal data.
  func logDebug(_ message: @autoclosure () -> String) {
    if debugEnabled {
      Logger.debug(message(), category: "admob")
    }
  }
}

/// Exported C symbol registered by the Rust core through
/// `register_ios_plugin(init_plugin_admob)` (`tauri::ios_plugin_binding!`).
@_cdecl("init_plugin_admob")
func initPlugin() -> Plugin {
  return AdmobPlugin()
}
